contracts/Everlong.sol

```diff
@@ -66,64 +67,313 @@ import { EVERLONG_KIND, EVERLONG_VERSION } from "./libraries/Constants.sol";
-        _decimals = success ? result : _DEFAULT_UNDERLYING_DECIMALS;
+    /// @notice Allows admin to transfer the admin role.
+    /// @param __admin The new admin address.
+    function setAdmin(address __admin) external onlyAdmin {
+        _admin = __admin;
+        emit AdminUpdated(__admin);
+    }
+
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ ERC4626                                                 │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /// @notice Calculate the total amount of assets controlled by everlong.
+    /// @notice To do this efficiently, the weighted average maturity is used.
+    /// @dev Underestimates the actual value by overestimating the average
+    ///      maturity of the portfolio.
+    /// @return Total amount of assets controlled by Everlong.
+    function totalAssets() public view override returns (uint256) {
+        // If everlong holds no bonds, return the balance.
+        uint256 balance = _asset.balanceOf(address(this));
+        if (_portfolio.totalBonds == 0) {
+            return balance;
+        }
+
+        // Estimate the value of everlong-controlled positions by calculating
+        // the proceeds one would receive from closing a position with the portfolio's
+        // total amount of bonds and weighted average maturity.
+        // The weighted average maturity is rounded to the next checkpoint
+        // timestamp to underestimate the value.
+        return
+            balance +
+            _hyperdrive.previewCloseLong(
+                _asBase,
+                IEverlong.Position({
+                    maturityTime: _hyperdrive
+                        .getCheckpointIdUp(_portfolio.avgMaturityTime)
+                        .toUint128(),
+                    bondAmount: _portfolio.totalBonds
+                })
+            );
+    }
+
+    /// @dev Frees sufficient assets for a withdrawal by closing positions.
+    /// @param _assets Amount of assets owed to the withdrawer.
+    function _beforeWithdraw(
+        uint256 _assets,
+        uint256
+    ) internal virtual override {
+        // Close more positions until sufficient idle to process withdrawal.
+        _closePositions(_assets - _asset.balanceOf(address(this)));
+    }
+
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ Rebalancing                                             │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /// @notice Rebalance the everlong portfolio by closing mature positions
+    ///         and using the proceeds to open new positions.
+    function rebalance() public override {
+        // Close matured positions.
+        _closeMaturedPositions();
+
+        // Spend idle on opening a new position.
+        uint256 toSpend = _asset.balanceOf(address(this));
+        IERC20(_asset).approve(address(_hyperdrive), toSpend);
```

Comment:
It would be good to use `forceApprove` here since it's possible for the asset to be a non-standard ERC20 token. It would also be good to add an additional 1 wei of approval so that the storage slot remains warm after the funds are pulled.